src/backend/core/models.py

```diff
@@ -867,32 +867,33 @@ def send_invitation_email(self, email, role, sender, language=None):
 
         self.send_email(subject, [email], context, language)
 
-    def delete(self, *args, **kwargs):
-        """Invalidate cache for number of accesses when deleting a document."""
-        super().delete(*args, **kwargs)
-        self.invalidate_nb_accesses_cache()
-
     @transaction.atomic
     def soft_delete(self):
         """
         Soft delete the document, marking the deletion on descendants.
         We still keep the .delete() method untouched for programmatic purposes.
         """
-        if self.deleted_at or self.ancestors_deleted_at:
-            raise RuntimeError(
-                "This document is already deleted or has deleted ancestors."
-            )
-
-        # Check if any ancestors are deleted
-        if self.get_ancestors().filter(deleted_at__isnull=False).exists():
+        if (
+            self._meta.model.objects.filter(
+                models.Q(deleted_at__isnull=False)
+                | models.Q(ancestors_deleted_at__isnull=False),
+                pk=self.pk,
+            ).exists()
+            or self.get_ancestors().filter(deleted_at__isnull=False).exists()
+        ):
             raise RuntimeError(
-                "Cannot delete this document because one or more ancestors are already deleted."
+                _("This document is already deleted or has deleted ancestors.")
             )
 
         self.ancestors_deleted_at = self.deleted_at = timezone.now()
         self.save()
         self.invalidate_nb_accesses_cache()
 
+        if parent := self.get_parent():
+            self._meta.model.objects.filter(pk=parent.pk).update(
```

Comment:
Why using `self._meta.model.objects` is better than using `Document.objects` ?

Reply:
I like that it can be reached from the args that the method gets and not from something external. Sometimes importing models globally leads to circular imports so I tend to limit them when I can.